src/controller/proxy/controller.go

```diff
@@ -0,0 +1,241 @@
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/controller/artifact"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/registry"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	maxWait             = 10
+	manifestListWaitSec = 900
+	sleepIntervalSec    = 20
+)
+
+var (
+	// Ctl is a global proxy controller instance
+	ctl  Controller
+	once sync.Once
+)
+
+// Controller defines the operations related with pull through proxy
+type Controller interface {
+	// UseLocalBlob check if the blob should use local copy
+	UseLocalBlob(ctx context.Context, art lib.ArtifactInfo) bool
+	// UseLocalManifest check manifest should use local copy
+	UseLocalManifest(ctx context.Context, art lib.ArtifactInfo) bool
+	// ProxyBlob proxy the blob request to the remote server
+	ProxyBlob(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error
+	// ProxyManifest proxy the manifest request to the remote server
+	ProxyManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error
+}
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	artifactCtl artifact.Controller
+	local       localInterface
+}
+
+// ControllerInstance -- Get the proxy controller instance
+func ControllerInstance() (Controller, error) {
+	// Lazy load the controller
+	// Because LocalHelper is not ready unless core startup completely
+	if ctl == nil {
+		once.Do(func() {
+			ctl = &controller{
+				blobCtl:     blob.Ctl,
+				registryMgr: registry.NewDefaultManager(),
+				artifactCtl: artifact.Ctl,
+				local:       newLocalHelper(),
+			}
+		})
+	}
+	return ctl, nil
+}
+
+func (c *controller) UseLocalBlob(ctx context.Context, art lib.ArtifactInfo) bool {
+	if len(art.Digest) > 0 {
+		exist, err := c.local.BlobExist(ctx, art)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, art lib.ArtifactInfo) bool {
+	if len(art.Digest) > 0 {
+		exist, err := c.local.ManifestExist(ctx, art)
+		if err != nil {
+			return false
+		}
+		return exist
+	}
+	return false
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error {
+	var man distribution.Manifest
+	var err error
+	remoteRepo := GetRemoteRepo(art)
+	r := newRemoteHelper(p.RegistryID)
+	ref := art.Digest
+	if len(ref) == 0 {
+		ref = art.Tag
+	}
+	man, err = r.Manifest(remoteRepo, ref)
+	if err != nil {
+		if errors.IsNotFoundErr(err) {
+			go func() {
+				c.local.DeleteManifest(remoteRepo, art.Tag)
```

Comment:
log errors？

Reply:
the art.Tag will not be changed so it can be used in enclosure
error is already logged in DeleteManifest.